text/0000-is-sorted.md

```diff
@@ -0,0 +1,272 @@
+`v.is_sorted()`, one would need to write `v.iter().is_sorted()`.
+
+This always works for `is_sorted()` because of the `Ord` blanket impl which
+implements `Ord` for all references to an `Ord` type. For `is_sorted_by` and
+`is_sorted_by_key` it would introduce an additional reference to the closures'
+arguments (i.e. `v.iter().is_sorted_by_key(|x| ...))` where `x` is `&&T`).
+
+While these two inconveniences are not deal-breakers, being able to call those
+three methods on slices (and all `Deref<Target=[T]>` types) directly, could be
+favourable for many programmers (especially given the popularity of slice-like
+data structures, like `Vec<T>`). Additionally, the `sort` method and friends
+are defined for slices, thus one might expect the `is_sorted()` method there,
+too.
+
+
+### Add the three methods to additional data structures (like `LinkedList`) as well
+Adding these methods to every data structure in the standard libary is a lot of
+duplicate code. Optimally, we would have a trait that represents sequential
+data structures and would only add `is_sorted` and friends to said trait. We
+don't have such a trait as of now; so `Iterator` is the next best thing. Slices
+deserve special treatment due to the reasons mentioned above (popularity and
+`sort()`).
+
+
+### `Iterator::while_sorted`, `is_sorted_until`, `sorted_prefix`, `num_sorted`, ...
+[In the issue on the main repository](https://github.com/rust-lang/rust/issues/44370),
+concerns about completely consuming the iterator were raised. Some alternatives,
+such as [`while_sorted`](https://github.com/rust-lang/rust/issues/44370#issuecomment-327873139),
+were suggested. However, consuming the iterator is neither uncommon nor a
+problem. Methods like `count()`, `max()` and many more consume the iterator,
+too. [One comment](https://github.com/rust-lang/rust/issues/44370#issuecomment-344516366) mentions:
+
+> I am a bit skeptical of the equivalent on Iterator just because the return
+> value does not seem actionable -- you aren't going to "sort" the iterator
+> after you find out it is not already sorted. What are some use cases for this
+> in real code that does not involve iterating over a slice?
+
+As mentioned above, `Iterator` is the next best thing to a trait representing
+sequential data structures. So to check if a `LinkedList`, `VecDeque` or
+another sequential data structure is sorted, one would simply call
+`collection.iter().is_sorted()`. It's likely that this is the main usage for
+`Iterator`'s `is_sorted` methods. Additionally, code like
+`if v.is_sorted() { v.sort(); }` is not very useful:  `sort()` already runs in
+O(n) for already sorted arrays.
+
+Suggestions like `is_sorted_until` are not really useful either: one can easily
+get a subslice or a part of an iterator (via `.take()`) and call `is_sorted()`
+on that part.
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+
+### Is `Iterator::is_sorted_by_key` useless?
```

Comment:
Yes, I think so, precisely for the reason you mention.

We already have quite many methods in `Iterator` and I think that additions should only be made if they increase readability and helps you express things which are quite tedious and unidiomatic to express otherwise. In this case I think `iter.map(extract).is_sorted()` is better than `iter.is_sorted_by_key(extract)`.

If there is popular demand for this later we can always add it then, but for now I think it is premature.